Firmware/mmu2_protocol_logic.cpp

```diff
@@ -128,256 +124,253 @@ void ProtocolLogic::SendMsg(RequestMsg rq) {
-            SendVersion(0);
-        } else {
-            logic->mmuFwVersionMajor = logic->rsp.paramValue;
-            if (logic->mmuFwVersionMajor != supportedMmuFWVersionMajor) {
-                if( --retries == 0){
-                    // if (--retries == 0) has a specific meaning - since we are losing bytes on the UART for no obvious reason
-                    // it can happen, that the reported version number is not complete - i.e. "1" instead of "19"
-                    // Therefore we drop the MMU only if we run out of retries for this very reason.
-                    // There is a limited amount of retries per the whole start seq.
-                    // We also must be able to actually detect an unsupported MMU FW version, so the amount of retries shall be kept small.
-                    return VersionMismatch;
-                } else {
-                    SendVersion(0);
-                }
+void ProtocolLogic::IdleRestart() {
+    scopeState = ScopeState::Ready;
+}
+
+StepStatus ProtocolLogic::ProcessVersionResponse(uint8_t stage) {
+    if (rsp.request.code != RequestMsgCodes::Version || rsp.request.value != stage) {
+        // got a response to something else - protocol corruption probably, repeat the query OR restart the comm by issuing S0?
+        SendVersion(stage);
+    } else {
+        mmuFwVersion[stage] = rsp.paramValue;
+        if (mmuFwVersion[stage] != pgm_read_byte(supportedMmuFWVersion + stage)) {
+            if (--retries == 0) {
+                return VersionMismatch;
             } else {
-                logic->dataTO.Reset(); // got meaningful response from the MMU, stop data layer timeout tracking
-                SendVersion(1);
+                SendVersion(stage);
             }
-        }
-        break;
-    case State::S1Sent: // received response to S1 - minor
-        if( logic->rsp.request.code != RequestMsgCodes::Version || logic->rsp.request.value != 1 ){
-            // got a response to something else - protocol corruption probably, repeat the query OR restart the comm by issuing S0?
-            SendVersion(1);
         } else {
-            logic->mmuFwVersionMinor = logic->rsp.paramValue;
-            if (logic->mmuFwVersionMinor != supportedMmuFWVersionMinor){
-                if( --retries == 0) {
-                    return VersionMismatch;
-                } else {
-                    SendVersion(1);
-                }
-            } else {
-                SendVersion(2);
-            }
+            dataTO.Reset(); // got a meaningful response from the MMU, stop data layer timeout tracking
+            SendVersion(stage + 1);
         }
-        break;
-    case State::S2Sent: // received response to S2 - revision
-        if( logic->rsp.request.code != RequestMsgCodes::Version || logic->rsp.request.value != 2 ){
+    }
+    return Processing;
+}
+
+StepStatus ProtocolLogic::ScopeStep() {
+    if ((uint_fast8_t)scopeState & (uint8_t)ScopeState::NotExpectsResponse) {
```

Comment:
I think this can be `if( !ExpectsResponse() )`, doesn't seem to save space but is more readable.